### DISCUSSION TITLE
[SDK] <Title> Javascript SDK: Unable to use testnet paymaster contract, always paid by ETH

### DISCUSSION THREAD

**Discussion by enochSatis** [zkSync-Community-Hub/zksync-developers#150] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

## Background setting

I am following the tutorial for using the testnet paymaster with Javascript SDK: https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html#paying-fees-using-testnet-paymaster

And also the example codes on gitHub: https://github.com/matter-labs/paymaster-examples/blob/main/contracts/test/erc20fixed.test.ts

I have deployed the demo `Greeter.sol` at `0x88Be52128A0CE7ed65c845c888bf2FFD69Aa271c`, and I would like to trigger the contract with the testnet paymaster. 

## My procedure

1. I get the testnet paymaster address with: `provider.getTestnetPaymasterAddress()`. The return is `0x8f0ea1312da29f17eabeb2f484fd3c112cccdd63` which seems correct.
2. I choose to pay with the testnet DAI, so I picked the token address `0x3e7676937A7E96CFB7616f255b9AD9FF47363D4b`, which should also be correct.
3. I set the paymaster estimation parameters according to the tutorial:
```
const paymasterEstimateParams = utils.getPaymasterParams(paymasterAddress, {
    type: "ApprovalBased",
    token: "0x3e7676937A7E96CFB7616f255b9AD9FF47363D4b",
    minimalAllowance: ethers.BigNumber.from(1),
    innerInput: new Uint8Array(),
});
```
4. Then I setup the paymaster parameters according to the tutorial:
```
const fee = gasPrice.mul(gasLimit.toString());
const paymasterParams = utils.getPaymasterParams(paymasterAddress, {
    type: "ApprovalBased",
    token: "0x3e7676937A7E96CFB7616f255b9AD9FF47363D4b",
    minimalAllowance: fee,
    // empty bytes as testnet paymaster does not use innerInput
    innerInput: new Uint8Array(),
});
```
5. Finally plugging the paymaster parameters into the contract interaction transaction:
```
const tx = await contract.setGreeting(newMessage, {
    maxFeePerGas: gasPrice,
    maxPriorityFeePerGas: ethers.BigNumber.from(0),
    gasLimit: gasLimit,
    customData: {
      gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
      paymasterParams
    },
});
```

## Problem
No matter how I tried, the transaction will success but the gas is always paid with ETH.

### Expected Behavior

I tried the following modifications, but the transaction will always be paid with ETH:

1. Try using paymaster with and without approving the token to the testnet paymaster contract.
2. Toggling the `paymasterParams` field in the `customData`:
```
customData: {
    gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
    paymasterParams,
    // paymasterParams: paymasterParams,
},
```
3. Using the `paymasterEstimationParams` (with `minimalAllowance: ethers.BigNumber.from(1)` as some other examples suggest)

### Code Example

My code is a modification of the demo `use-greeter.ts`. I have hardcoded the addresses except for the wallet private key, so anyone can run this directly.

# use-greeter.ts

```
import { Provider, utils } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";

// load env file
import dotenv from "dotenv";
dotenv.config();

// load contract artifact. Make sure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.sol/Greeter.json";

const PRIVATE_KEY = process.env.WALLET_PRIVATE_KEY || "";

// Pay in DAI on testnet
const PAYMASTER_TOKEN_ADDR = "0x3e7676937A7E96CFB7616f255b9AD9FF47363D4b";
console.log(`Paymaster token address: ${PAYMASTER_TOKEN_ADDR}`);

if (!PRIVATE_KEY)
  throw "⛔️ Private key not detected! Add it to the .env file!";

// Address of the contract on zksync testnet
const CONTRACT_ADDRESS = "0x88Be52128A0CE7ed65c845c888bf2FFD69Aa271c";

if (!CONTRACT_ADDRESS) throw "⛔️ Contract address not provided";

// An example of a deploy script that will deploy and call a simple contract.
export default async function (hre: HardhatRuntimeEnvironment) {

  console.log(`Running script to interact with contract ${CONTRACT_ADDRESS}`);

  // Initialize the provider.
  // @ts-ignore
  const provider = new Provider(hre.userConfig.networks?.zkSyncTestnet?.url);
  const signer = new ethers.Wallet(PRIVATE_KEY, provider);

  var paymasterAddress = await provider.getTestnetPaymasterAddress();
  paymasterAddress = String(paymasterAddress);
  console.log(`Testnet paymaster address: ${paymasterAddress}`);


  const paymasterEstimateParams = utils.getPaymasterParams(paymasterAddress, {
    type: "ApprovalBased",
    token: PAYMASTER_TOKEN_ADDR,
    minimalAllowance: ethers.BigNumber.from(1),
    innerInput: new Uint8Array(),
  });

  // Initialize contract instance
  const contract = new ethers.Contract(
    CONTRACT_ADDRESS,
    ContractArtifact.abi,
    signer
  );

  // Read message from contract
  console.log(`The message is ${await contract.greet()}`);

  // send transaction to update the message
  const newMessage = "Greetings Programs!";

  const gasPrice = await provider.getGasPrice();
  const gasLimit = await contract.estimateGas.setGreeting(newMessage, {
    customData: {
      gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
      // paymasterParams: paymasterEstimateParams,
      paymasterEstimateParams,
    },
  });
  console.log(`Gas price: ${gasPrice}`);
  console.log(`Gas limit: ${gasLimit}`);

  const fee = gasPrice.mul(gasLimit.toString());
  console.log(`Fee: ${fee}`);



  const paymasterParams = utils.getPaymasterParams(paymasterAddress, {
    type: "ApprovalBased",
    token: PAYMASTER_TOKEN_ADDR,
    minimalAllowance: fee,
    // empty bytes as testnet paymaster does not use innerInput
    innerInput: new Uint8Array(),
  });

  const tx = await contract.setGreeting(newMessage, {
    maxFeePerGas: gasPrice,
    maxPriorityFeePerGas: ethers.BigNumber.from(0),
    gasLimit: gasLimit,
    customData: {
      gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
      // paymasterParams: paymasterParams,
      paymasterParams
    },
  });
  console.log(`Transaction to change the message is ${tx.hash}`);
  await tx.wait();

  // Read message after transaction
  console.log(`The message now is ${await contract.greet()}`);
}
```

# package.json
```
{
  "name": "zksync-hardhat-template",
  "version": "1.0.0",
  "main": "index.js",
  "author": "Antonio <aug@matterlabs.dev>",
  "license": "MIT",
  "devDependencies": {
    "@matterlabs/hardhat-zksync-deploy": "^0.6.5",
    "@matterlabs/hardhat-zksync-solc": "^0.4.2",
    "@matterlabs/hardhat-zksync-verify": "^0.1.8",
    "@matterlabs/zksync-contracts": "^0.6.1",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "@openzeppelin/contracts": "^4.9.3",
    "@types/chai": "^4.3.4",
    "@types/mocha": "^10.0.1",
    "chai": "^4.3.7",
    "dotenv": "^16.0.3",
    "ethers": "^5.7.2",
    "hardhat": "^2.12.4",
    "mocha": "^10.2.0",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.4",
    "zksync-web3": "^0.14.3"
  },
  "scripts": {
    "test": "NODE_ENV=test hardhat test --network zkSyncTestnet",
    "deploy": "yarn hardhat deploy-zksync --script deploy-greeter.ts",
    "greet": "yarn hardhat deploy-zksync --script use-greeter.ts"
  }
}
```

**Comment by dutterbutter** [comment]
Thanks for the very well documented issue. The setup looks correct at first glance, what rpc are you connecting to in your hardhat config (e.g. `hre.userConfig.networks?.zkSyncTestnet?.url`)? Also what wallet are you using?

**Reply by enochSatis** [reply]
Hi @dutterbutter.

For zkSync testnet URL, I am using:
```
url: "https://zksync2-testnet.zksync.dev"
```

I think the RPC URL is working fine? At least I can deploy contracts and interact with them. 

For the wallet, I am just using a simple EOA wallet generated from Metamask. Here is my wallet address if you are interested:
```
0x512ae84232e12145ad7bf3dBd7862Ea75F214D77
```

**Reply by dutterbutter** [reply]
I was able to reproduce on my side, will investigate further.

**Reply by enochSatis** [reply]
Thanks 🙏

**Reply by enochSatis** [reply]
@dutterbutter Hi is there any update on this issue?

**Reply by dutterbutter** [reply]
Hey @enochSatis I can't seem to determine the root cause here. The [Hello world example](https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html) that makes use of the testnet paymaster works fine and successfully pays fees using DAI. I have deployed other paymasters to testnet using the approval based flow to pay fees in DAI and USDC and that works fine as well. 

We do intend to deprecate the testnet paymaster as it was an effort to showcase the capabilities initially. I apologize for the inconvenience here and will follow up if I do discover the root cause.

**Reply by enochSatis** [reply]
Got it. I will try to deploy my own paymaster contract and try again. Thanks for your help.

**Reply by dutterbutter** [reply, marked as answer]
@enochSatis actually I see the issue! I was able to successfully run this and pay with DAI. 

The issue is you are using `ethers.Wallet` rather than the our sdks `Wallet`. So try this:

```
import { Provider, utils, Wallet } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";

// load env file
import dotenv from "dotenv";
dotenv.config();

// load contract artifact. Make sure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.sol/Greeter.json";

const PRIVATE_KEY = process.env.WALLET_PRIV_KEY || "";

// Pay in DAI on testnet
const PAYMASTER_TOKEN_ADDR = "0x3e7676937A7E96CFB7616f255b9AD9FF47363D4b";
console.log(`Paymaster token address: ${PAYMASTER_TOKEN_ADDR}`);

if (!PRIVATE_KEY)
  throw "⛔️ Private key not detected! Add it to the .env file!";

// Address of the contract on zksync testnet
const CONTRACT_ADDRESS = "0x88Be52128A0CE7ed65c845c888bf2FFD69Aa271c";

if (!CONTRACT_ADDRESS) throw "⛔️ Contract address not provided";

// An example of a deploy script that will deploy and call a simple contract.
export default async function (hre: HardhatRuntimeEnvironment) {

  console.log(`Running script to interact with contract ${CONTRACT_ADDRESS}`);

  // Initialize the provider.
  // @ts-ignore
  const provider = new Provider(hre.userConfig.networks?.zkSyncTestnet?.url);
  const signer = new Wallet(PRIVATE_KEY, provider);

  var paymasterAddress = await provider.getTestnetPaymasterAddress();
  paymasterAddress = String(paymasterAddress);
  console.log(`Testnet paymaster address: ${paymasterAddress}`);


  const paymasterEstimateParams = utils.getPaymasterParams(paymasterAddress, {
    type: "ApprovalBased",
    token: PAYMASTER_TOKEN_ADDR,
    minimalAllowance: ethers.BigNumber.from(1),
    innerInput: new Uint8Array(),
  });

  // Initialize contract instance
  const contract = new ethers.Contract(
    CONTRACT_ADDRESS,
    ContractArtifact.abi,
    signer
  );

  // Read message from contract
  console.log(`The message is ${await contract.greet()}`);

  // send transaction to update the message
  const newMessage = "Greetings Programs!";

  const gasPrice = await provider.getGasPrice();
  const gasLimit = await contract.estimateGas.setGreeting(newMessage, {
    customData: {
      gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
      // paymasterParams: paymasterEstimateParams,
      paymasterEstimateParams,
    },
  });
  console.log(`Gas price: ${gasPrice}`);
  console.log(`Gas limit: ${gasLimit}`);

  const fee = gasPrice.mul(gasLimit.toString());
  console.log(`Fee: ${fee}`);

  const paymasterParams = utils.getPaymasterParams(paymasterAddress, {
    type: "ApprovalBased",
    token: PAYMASTER_TOKEN_ADDR,
    minimalAllowance: fee,
    // empty bytes as testnet paymaster does not use innerInput
    innerInput: new Uint8Array(),
  });

  const tx = await contract.connect(signer).setGreeting(newMessage, {
    maxFeePerGas: gasPrice,
    maxPriorityFeePerGas: ethers.BigNumber.from(0),
    gasLimit: gasLimit,
    customData: {
      gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
      // paymasterParams: paymasterParams,
      paymasterParams
    },
  });
  console.log(`Transaction to change the message is ${tx.hash}`);
  await tx.wait();

  // Read message after transaction
  console.log(`The message now is ${await contract.greet()}`);
}
```

And let me know if that works for you. You will notice we are using `Wallet` from `zksync-web3` and connecting to when doing the `setGreeter` call.

**Reply by enochSatis** [reply]
I see! Thanks for your help!

**Comment by augustinov** [comment]
It seems that you are setting up the paymaster parameters correctly, including the token address and minimal allowance. However, it's important to note that the paymaster logic needs to be implemented on the server-side as well.

Make sure that the paymaster contract logic is properly configured to accept payments in the specified token and process transactions accordingly. Additionally, ensure that the paymaster contract is properly funded with the required token balance to cover the transaction fees.

If you've verified the server-side implementation of the paymaster contract and the issue persists, consider reaching out to the zkSync support channels or community forums for further assistance and troubleshooting. They may provide insights or guidance on resolving the issue with paymaster transactions not being paid with the expected token.

**Comment by dawn71v** [comment]
Ensure that the paymasterParams are correctly configured and passed into the contract interaction transaction. Check the parameters such as the token address, minimal allowance, and inner input to ensure they match the requirements of the testnet paymaster. Verify that the gas limit and gas price are accurately calculated and set in the transaction object. Additionally, confirm that the testnet paymaster contract is properly configured to handle token transfers and approvals.

**Comment by dunminecom** [comment]
It seems that the issue might be related to how the paymaster parameters are configured or utilized in the transaction. Double-checking the configuration and ensuring that the paymaster parameters are correctly set and passed to the transaction might help resolve the problem. Additionally, reviewing the paymaster contract implementation and its interaction with the testnet DAI token could provide insights into any potential issues.
